src/main/kotlin/cacophony/automata/minimalization/DFAUtils.kt

```diff
@@ -0,0 +1,101 @@
+package cacophony.automata.minimalization
+
+import cacophony.automata.DFA
+
+class DFAPreimagesCalculator<DFAState>(
+    dfa: DFA<DFAState>,
+) {
+    private val cache: Map<DFAState, Map<Char, Set<DFAState>>> =
+        dfa
+            .getProductions()
+            .entries
+            .groupBy({ it.value }, { it.key })
+            .mapValues { entry ->
+                entry.value.groupBy({ it.second }, { it.first }).mapValues { it.value.toSet() }
+            }
+
+    fun getPreimages(states: Set<DFAState>): Map<Char, Set<DFAState>> =
+        states
+            .mapNotNull { state -> cache[state]?.let { state to it } }
+            .flatMap { it.second.entries }
+            .flatMap { entry -> entry.value.map { entry.key to it } }
+            .groupBy({ it.first }, { it.second })
+            .mapValues { it.value.toSet() }
+}
+
+fun <DFAState> getDFAForEmptyLanguage(): DFA<DFAState> = TODO("DFA interface does not allow empty languages")
```

Comment:
How it doesn't allow it? Isn't it enough to create DFA without accepting states?

Reply:
Maybe I misunderstood the spec, but I thought that non-null dead states are not allowed? This is why I asked to make DFA.getStartingState() return DFAState? instead of DFAState.